buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/incubation/IncubatingApiReportTask.kt

```diff
@@ -225,3 +206,169 @@ class Analyzer @Inject constructor(
+
+    override fun collectFrom(sourceFile: File): VersionsToIncubating {
+        val versionsToIncubating = mutableMapOf<Version, MutableSet<IncubatingDescription>>()
+        if (!sourceFile.name.endsWith(".java")) {
+            return versionsToIncubating
+        }
+        JavaParser.parse(sourceFile).run {
+            solver.inject(this)
+            findAllIncubating()
+                .map { node -> toVersionIncubating(sourceFile, node) }
+                .forEach { (version, incubating) ->
+                    versionsToIncubating.getOrPut(version) { mutableSetOf() }.add(incubating)
+                }
+        }
+        return versionsToIncubating
+    }
+
+    private
+    fun CompilationUnit.findAllIncubating() =
+        findAll(Node::class.java).filter { it.isIncubating }
+
+    private
+    val Node.isIncubating: Boolean
+        get() = (this as? NodeWithAnnotations<*>)?.annotations?.any { it.nameAsString == "Incubating" } ?: false
+
+    private
+    fun CompilationUnit.toVersionIncubating(sourceFile: File, node: Node) =
+        Pair(
+            findVersionFromJavadoc(node as NodeWithAnnotations<*>) ?: versionNotFound,
+            nodeName(node, this, sourceFile)
+        )
+
+    private
+    fun findVersionFromJavadoc(node: NodeWithAnnotations<*>): String? = if (node is NodeWithJavadoc<*>) {
+        node.javadoc.map { javadoc ->
+            javadoc.blockTags
+                .find { tag -> tag.tagName == "since" }
+                ?.content?.elements?.find { description -> description is JavadocSnippet }
+                ?.toText()
+        }.orElse(null)
+    } else {
+        null
+    }
+
+    private
+    fun nodeName(it: Node?, unit: CompilationUnit, file: File) = when (it) {
+        is EnumDeclaration -> tryResolve({ it.resolve().qualifiedName }) { inferClassName(unit) }
+        is ClassOrInterfaceDeclaration -> tryResolve({ it.resolve().qualifiedName }) { inferClassName(unit) }
+        is MethodDeclaration -> tryResolve({ it.resolve().qualifiedSignature }) { inferClassName(unit) }
+        is AnnotationDeclaration -> tryResolve({ it.resolve().qualifiedName }) { inferClassName(unit) }
+        is NodeWithSimpleName<*> -> it.nameAsString
+        else -> unit.primaryTypeName.orElse(file.name)
+    }
+
+    private
+    fun inferClassName(unit: CompilationUnit) =
+        "${unit.packageDeclaration.map { it.nameAsString }.orElse("")}.${unit.primaryTypeName.orElse("")}"
+
+    private
+    inline fun tryResolve(resolver: () -> String, or: () -> String) = try {
+        resolver()
+    } catch (e: Throwable) {
+        or()
+    }
+}
+
+
+private
+class KotlinVersionsToIncubatingCollector(srcDir: File) : VersionsToIncubatingCollector {
+
+    override fun collectFrom(sourceFile: File): VersionsToIncubating {
+
+        val versionsToIncubating = mutableMapOf<Version, MutableSet<IncubatingDescription>>()
+
+        if (!sourceFile.name.endsWith(".kt")) {
```

Comment:
⭕️ I'd move this to the top of the method and return `emptyMap()` to remove the dependency on the variable and clarify it as a precondition